src/java.desktop/share/classes/javax/swing/text/DefaultEditorKit.java

```diff
@@ -339,7 +339,13 @@ public void write(Writer out, Document doc, int pos, int len)
                 for (int counter = last; counter < maxCounter; counter++) {
                     if (array[counter] == '\n') {
                         if (counter > last) {
-                            out.write(array, last, counter - last);
+                            if(array[counter-1] == '\r') {
+                                out.write(array, last, counter - last - 1);
+                            }
+                            else
+                            {
+                                out.write(array, last, counter - last);
+                            }
                         }
                         out.write(endOfLine);
```

Comment:
Yes, Write function copies string without endOfLine, and endOfLine is appended to it after the whole string is copied. 
Before also it was copying string without endOfLine which included '\r', now this logic copies string characters before '\r' and then appends endOfLine.

Reply:
ok, so you are telling out.write(array, last, counter - last - 1) will write till \r\n and out.write(endOfLine) will write \r\n, is it?
ANyway, you need to modify this change as per coding style which is 
```
else {
}
```

Reply:
> ok, so you are telling out.write(array, last, counter - last - 1) will write till \r\n and out.write(endOfLine) will write \r\n, is it?

Yes.

Reply:
So, presently it was writing <String>/r/n/r/n and with fix, it will write <String>/r/n, right?

Reply:
No, presently it was writing /r/r/n and with fix it will write /r/n for every line ending...... /r was causing for carriage return and /r/n for an extra empty line on Text pane.......

Reply:
But should /r/r/n cause 2 line feeds? /r should be just carriage return to cause caret to move from last column to first, why it is inserting new line?

Reply:
It is common to treat `\r` as new line if it's not followed by `\n`. In a visual editor `\r` doesn't do anything.

Yet it's an interesting detail that wasn't covered in the initial evaluation.

Reply:
> But should /r/r/n cause 2 line feeds? /r should be just carriage return to cause caret to move from last column to first, why it is inserting new line?

\r is converted to \n in setText method....... To be specific, in `read(Reader in, Document doc, int pos)` method of DefaultEditorKit class......

Reply:
So, that means presently it is writing /n/r/n not /r/r/n, right?

Reply:
Initially it is /r/r/n from `getText()` of pane0, later it is /n/r/n from `setText()`....

Reply:
When you do out.write(array, last, counter - last); in write(Writer out, Document doc, int pos, int len) method, what is being written? Is it /r/r/n or /n/r/n presently?

Reply:
Finally its /n/r/n....... I mentioned /r/r/n because that was the output string from `getText()` method to `setText()` method......